[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in SocketConnection::sendMessage()

aperezdc opened the pull request:
#### eb7d8f7e49e20db63f127cb4dfd447c9f5106105
<pre>
[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in SocketConnection::sendMessage()
<a href="https://bugs.webkit.org/show_bug.cgi?id=302780">https://bugs.webkit.org/show_bug.cgi?id=302780</a>

Reviewed by Michael Catanzaro.

Rewrite SocketConnection::sendMessage() to use spans and their helper
functions to prepare messages for sending. As a bonus, its structure now
follows that of ::readMessage(), making it easier to understand both as
a whole.

* Source/WTF/wtf/glib/SocketConnection.cpp:
(WTF::SocketConnection::readMessage): Hoist {Mini,Maxi}mumMessageBodySize to
global scope, to reuse the definitions in SocketConnection::sendMessage().
(WTF::SocketConnection::sendMessage):
* Source/WTF/wtf/glib/SocketConnection.h: Change parameter declaration to
use CString instead of a raw string pointer.

Canonical link: <a href="https://commits.webkit.org/303283@main">https://commits.webkit.org/303283@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a78e4eab34dfd633a0ee484cdd6c987c5c225a76

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131933 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/4425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42945 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139445 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4365 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/4187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/100847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134879 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/118154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/81638 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82665 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/123996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142090 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130440 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/4095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/36857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/4176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/4187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109385 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27702 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/114434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/4148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/32832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163407 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/3980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/67595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/42484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/4108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/glib/SocketConnection.cpp:
```diff
@@ -188,45 +188,41 @@ bool SocketConnection::readMessage()
+        ASSERT(parametersSize == messageData.size());
+        memcpySpan(messageData, span(parameters));
+    }
 
```
I think here you could add `ASSERT(messageData.size() == 0)` to make sure the size is correct, right? (You would have to also change `memcpySpan(messageData, span(parameters));` to use `consumeSpan()`.)

aperezdc replied:
If the goal is to have the assertion cover the case of `parameters` being null, then we can just move the assertion outside of the “if” block, because `parametersSize` will be zero in that case. It would look like this, which looks clearer to me:

```cpp
    ASSERT(parametersSize == messageData.size());
    if (parameters)
        memcpySpan(messageData, span(parameters));
```

WDYT?

mcatanzaro replied:
That looks good.

aperezdc replied:
Let's apply this while landing, thanks for reviewing!

mcatanzaro reviewed (approved):
Thanks for fixing this. Getting rid of unsafe buffer usage is great.

I see this works because `consumeAndReinterpretCastTo` effectively returns the original span, from before the consume. Seems confusing to me, but OK.

I wonder why we have to use `htonl()` if we also have a flag to specify the byte order. Odd.

aperezdc committed:
[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in SocketConnection::sendMessage()
https://bugs.webkit.org/show_bug.cgi?id=302780

Reviewed by Michael Catanzaro.

Rewrite SocketConnection::sendMessage() to use spans and their helper
functions to prepare messages for sending. As a bonus, its structure now
follows that of ::readMessage(), making it easier to understand both as
a whole.

* Source/WTF/wtf/glib/SocketConnection.cpp:
(WTF::SocketConnection::readMessage): Hoist {Mini,Maxi}mumMessageBodySize to
global scope, to reuse the definitions in SocketConnection::sendMessage().
(WTF::SocketConnection::sendMessage):
* Source/WTF/wtf/glib/SocketConnection.h: Change parameter declaration to
use CString instead of a raw string pointer.

Canonical link: https://commits.webkit.org/303283@main
